packages/react-reconciler/src/ReactFiberCommitWork.new.js

```diff
@@ -892,7 +879,35 @@ function commitUnmount(
             const {destroy, tag} = effect;
             if (destroy !== undefined) {
               if ((tag & HookPassive) !== NoHookEffect) {
-                enqueuePendingPassiveHookEffectUnmount(current, effect);
+                effect.tag |= HookHasEffect;
+
+                // subtreeTags bubble in resetChildLanes which doens't get called for unmounted subtrees.
+                // So in the case of unmounts, we need to bubble passive effects explicitly.
```

Comment:
I was thinking we could move all this logic to the passive commit phase. We would have a separate subtree tag that represents whether there are mounted passive effects anywhere in the tree. It would always be set regardless of whether the dependencies changed, or whether there's an unmount callback.

That means we might visit some nodes that have effects but don't have destroy effects, but that's not a big deal.

We'd use this same flag in the Offscreen implementation to find effects that need to be mounted/unmounted.

Reply:
> I was thinking we could move all this logic to the passive commit phase.

Sorry, that was a confusing way to phrase my suggestion. I meant to say, we can avoid bubbling the `subtreeTag` here  by adding a new bit that is always set for any parent of a mounted effect.

Reply:
Just to make sure I'm understanding your suggestion correctly- the problem this loop is addressing is that effects deep inside of an unmounted tree aren't reflected by the `subtreeTag` value unless we recurse and find them. Setting a special bit for mounted effects wouldn't help– since we always reset `subtreeTag` to `NoSubtreeEffect` during cloning– unless we were to just never reset that special bit.

So the relevant line in `createWorkInProgress` would become:
```js
workInProgress.subtreeTag &= SpecialNewTagWeAreTalkingAbout;
```

Is this what you had in mind?

Reply:
Yeah that's right. We're planning a similar type of tag to track which trees contain portals. So it could be `&= StaticTags` or something like that, where `StaticTags` is the union of all these types of bits.

Reply:
Okay. This makes perfect sense then. I'll make the change.

Reply:
Although you need to reset the bit when all the children that have mounted effects are deleted. So I think this is more like the tree duration flags. You need a bit that represents "self", and a bit that represents the children.

Reply:
I guess the "self" tags go on `effectTag` field. Does what I'm saying make sense?

Reply:
I think 157aa81d7 is the change we discussed.

Reply:
Huh, your last two comments did not show up until after I pushed [157aa81](https://github.com/facebook/react/pull/19374/commits/157aa81d7b7fda4dd7225c474b20ef0214eb1d96) and commented.

Yes, `effectTag` is the equivalent of `selfBaseDuration` in this case.

> Although you need to reset the bit when all the children that have mounted effects are deleted. So I think this is more like the tree duration flags. You need a bit that represents "self", and a bit that represents the children.

I don't know how to achieve this without deeply traversing the entire subtree (not just the unmounted bit). How else would we know there were no more passive effects? After all, we aren't counting them.

Reply:
I believe that as of [this change](https://github.com/facebook/react/pull/19374/commits/157aa81d7b7fda4dd7225c474b20ef0214eb1d96) we have two potential over-traversal cases:
* Bailout, although we can avoid that by checking for it in `flushPassiveMountEffects`.
* After the last passive effect has been unmounted in a subtree, we might still traverse part of the subtree (since we never reset the `Static` effects).

Is this second over-traversal safe? I think it's what your previous comment is pointing out, but I don't see how your suggestion for addressing it would work.

Reply:
One potential issue without the bailout case is that Offscreen children also satisfy the condition for bailing out. I could check whether a fiber is in an Offscreen subtree but it seems slightly messy, so I wonder if there's a better way to determine whether or not children not being cloned is because of bailout or because of Offscreen (perhaps a tag on the fiber indicating that we should stop traversing?)

Reply:
We talked out of band and a summary is:
* We should add this new static bit to `effectTag` rather than `subtreeTag`.
* The static bit will then bubble to `subtreeTag` during complete phase for things that are still mounted.